lock-api/src/test/java/com/palantir/lock/client/ProfilingTimelockServiceTest.java

```diff
@@ -0,0 +1,236 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.client;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.when;
+
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.BooleanSupplier;
+
+import org.junit.Test;
+import org.mockito.ArgumentCaptor;
+import org.slf4j.Logger;
+
+import com.google.common.base.Stopwatch;
+import com.google.common.base.Ticker;
+import com.google.common.collect.ImmutableSet;
+import com.palantir.lock.StringLockDescriptor;
+import com.palantir.lock.v2.LockRequest;
+import com.palantir.lock.v2.TimelockService;
+import com.palantir.lock.v2.WaitForLocksRequest;
+import com.palantir.logsafe.Arg;
+
+public class ProfilingTimelockServiceTest {
+    private static final Duration SHORT_DURATION = ProfilingTimelockService.SLOW_THRESHOLD.dividedBy(5);
+    private static final Duration LONG_DURATION = ProfilingTimelockService.SLOW_THRESHOLD.multipliedBy(5);
+    private static final Duration TWO_CENTURIES = ChronoUnit.CENTURIES.getDuration().multipliedBy(2);
+
+    private final Logger logger = mock(Logger.class);
+    private final TimelockService delegate = mock(TimelockService.class);
+
+    private long clock = 0;
+    private long operationTiming = 0;
+    private final Ticker ticker = new Ticker() {
+        @Override
+        public long read() {
+            clock += operationTiming;
+            return clock;
+        }
+    };
+
+    private final AtomicBoolean allowLogging = new AtomicBoolean();
```

Comment:
I'm somewhat struggling to make sense of the tests. I'm not sure whether having methods like:

```java
flushLogsWithCall(LONG_DURATION, profilingTimelockService::getFreshTimestamp);
accumulateLogsWithCall(SHORT_DURATION, profilingTimelockService::getFreshTimestamp);
```

they'll still do the same things, I just think it's closer to the behaviour you're trying to achieve,

and assertions will be somewhat clearer i.e.

```java
accumulateLogsWithCall(...) // call 1
...
accumulateLogsWithCall(...) // call n-1
flushLogsWithCall(...) // call n
assert based on the slowest call between 1 and n
```